cmd/raidz_test/raidz_test.c

```diff
@@ -0,0 +1,800 @@
+/*
+ * Copyright (C) 2016 Gvozden Nešković. All rights reserved.
+ */
+
+#include <sys/zfs_context.h>
+#include <sys/time.h>
+#include <sys/wait.h>
+#include <sys/zio.h>
+#include <umem.h>
+#include <sys/vdev_raidz.h>
+#include <sys/vdev_raidz_impl.h>
+#include <assert.h>
+#include <stdio.h>
+#include "raidz_test.h"
+
+#define	D_INFO	1
+#define	D_DEBUG	2
+
+static int *rand_data;
+raidz_test_opts_t rto_opts;
+
+static char gdb[256];
+static const char gdb_tmpl[] = "gdb -ex \"set pagination 0\" -p %d";
+
+static void sig_handler(int signo)
+{
+	struct sigaction action;
+	/*
+	 * Restore default action and re-raise signal so SIGSEGV and
+	 * SIGABRT can trigger a core dump.
+	 */
+	action.sa_handler = SIG_DFL;
+	sigemptyset(&action.sa_mask);
+	action.sa_flags = 0;
+	(void) sigaction(signo, &action, NULL);
+
+	if (rto_opts.rto_gdb) {
+
+		if (system(gdb));
+	}
+
+	raise(signo);
+}
+
+static void print_opts(raidz_test_opts_t *opts, boolean_t force)
+{
+	char *verbose;
+	switch (opts->rto_v) {
+		case 0:
+			verbose = "no";
+			break;
+		case 1:
+			verbose = "info";
+			break;
+		default:
+			verbose = "debug";
+			break;
+	}
+
+	if (force || opts->rto_v >= D_INFO) {
+		PRINT(DBLSEP);
+		PRINT("Running with options:\n"
+		    "  (-a) zio ashift                   : %zu\n"
+		    "  (-o) zio offset                   : 1 << %zu\n"
+		    "  (-d) number of raidz data columns : %zu\n"
+		    "  (-s) size of DATA                 : 1 << %zu\n"
+		    "  (-S) sweep parameters             : %s \n"
+		    "  (-v) verbose                      : %s \n\n",
+		    opts->rto_ashift,			/* -a */
+		    ilog2(opts->rto_offset),		/* -o */
+		    opts->rto_dcols,			/* -d */
+		    ilog2(opts->rto_dsize),		/* -s */
+		    opts->rto_sweep ? "yes" : "no",	/* -S */
+		    verbose 				/* -v */
+		);
+	}
+}
+
+static void usage(boolean_t requested)
+{
+	const raidz_test_opts_t *o = &rto_opts_defaults;
+
+	FILE *fp = requested ? stdout : stderr;
+
+	(void) fprintf(fp, "Usage:\n"
+	"\t[-a zio ashift (default: %zu)]\n"
+	"\t[-o zio offset, exponent radix 2 (default: %zu)]\n"
+	"\t[-d number of raidz data columns (default: %zu)]\n"
+	"\t[-s zio size, exponent radix 2 (default: %zu)]\n"
+	"\t[-S parameter sweep (default: %s)]\n"
+	"\t[-t timeout for parameter sweep test]\n"
+	"\t[-B benchmark all raidz implementations]\n"
+	"\t[-v increase verbosity (default: %zu)]\n"
+	"\t[-h (print help)]\n"
+	"\t[-T test the test, see if failure would be detected]\n"
+	"\t[-D debug (attach gdb on SIGSEGV)]\n"
+	"",
+	o->rto_ashift,				/* -a */
+	ilog2(o->rto_offset),			/* -o */
+	o->rto_dcols,				/* -d */
+	ilog2(o->rto_dsize),			/* -s */
+	rto_opts.rto_sweep ? "yes" : "no",	/* -S */
+	o->rto_v				/* -d */
+	);
+
+	exit(requested ? 0 : 1);
+}
+
+static void process_options(int argc, char **argv)
+{
+	char *end;
+	size_t value;
+	int opt;
+
+	raidz_test_opts_t *o = &rto_opts;
+
+	bcopy(&rto_opts_defaults, o, sizeof (*o));
+
+	while ((opt = getopt(argc, argv,
+	    "TDBSvha:o:d:s:t:")) != EOF) {
```

Comment:
The `getopt` return code should be checked against -1 not EOF.   They're not always the same and the getopt man page explicitly says `-1` is the value on error.